File: src/c/ini_parser.c
/**
 * $File: ini_parser.c $
 * $Date: 2018-02-12 03:36:42 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2018 by Shen, Jen-Chieh $
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include <ini_parser.h>


 /**
  * @enum bool
  * @brief boolean data type.
  */
typedef enum bool {
	false,
	true,
} bool;


/* Check the line is a comment. */
static bool is_comment_line(char *line) {
	char* ch = line;

	while (*ch != '\0') {
		if (*ch != ' ' && *ch != '#')
			return false;

		if (*ch == '#')
			return true;

		ch++;
	}

	return false;
}

/* Get the property value by keyword. */
char* get_property(pair **in_data, int len, const char *keyword) {
	int index;

	for (index = 0;
		index < len;
		index++)
	{
		if (strcmp(in_data[index]->keyword, keyword) == 0)
			return in_data[index]->value;
	}

	return "";
}

/* Read .ini/.properties file. */
void read_ini_file(pair **in_data, const char *filePath) {
	FILE *fp;
	char line[STR_LEN];
	int data_len = 0;

	fp = fopen(filePath, "r");

	while (fgets(line, sizeof(line), fp)) {
		if (is_comment_line(line))
			continue;

		strcpy(in_data[data_len]->keyword, strtok(line, "="));
		strcpy(in_data[data_len]->value, strtok(NULL, "=\n"));

		data_len++;
	}

	fclose(fp);
}

File: src/js/ini-parser.js
/**
 * $File: ini-parser.js $
 * $Date: 2018-02-12 17:59:52 $
 * $Revision: $
 * $Creator: Jen-Chieh Shen $
 * $Notice: See LICENSE.txt for modification and distribution information
 *                   Copyright © 2018 by Shen, Jen-Chieh $
 */

"use strict";

/*
 * References:
 * @url https://gist.github.com/Arahnoid/9925725
 */


/**
 * Read a .ini/.properties file.
 *
 * TODO(jenchieh): currently only work in NodeJS environment.
 *
 * @param { string } filePath : file path to ini.
 */
function readINIFile(filePath) {
  // dictionary object.
  var ini_data = {};

  var txtFile = new File(filePath);
  txtFile.open("r");

  while (!txtFile.eof) {
    var line = txtFile.readln();

    if (isCommentLine(line))
      continue;

    var keyword = line.split('=')[0];
    var value = line.split('=')[1];
    ini_data[keyword] = value;
  }

  return ini_data;
}

/**
 * Check if line a comment line.
 * @param { string } line : line to check.
 * @return { bool } : true, is comment line. false, is data line.
 */
function isCommentLine(line) {
  for (var index = 0;
       index < line.length;
       ++index) {
    var ch = line.charAt(index);

    if (ch != ' ' && ch != '#')
      return false;

    if (ch == '#')
      return true;
  }

  return false;
}
